app/views/spree/amazon/address.html.erb

```diff
@@ -23,22 +23,58 @@
 
   <script>
     $( document ).ready(function() {
-      new OffAmazonPayments.Widgets.AddressBook({
-        sellerId: '<%= @amazon_gateway.preferred_merchant_id %>',
-        agreementType: 'orderreference',
+      <% unless current_order.amazon_transaction.try(:retry) %>
+        new OffAmazonPayments.Widgets.AddressBook({
+          sellerId: '<%= @amazon_gateway.preferred_merchant_id %>',
+          agreementType: 'orderreference',
           onOrderReferenceCreate: function(orderReference) {
-            order_reference = orderReference.getAmazonOrderReferenceId();
+              order_reference = orderReference.getAmazonOrderReferenceId();
+            },
+          onAddressSelect: function(orderReference) {
+            <%= render 'spree/amazon/payment' %>
           },
-        onAddressSelect: function(orderReference) {
-          <%= render 'spree/amazon/payment' %>
-        },
-        design: {
-          designMode: 'responsive'
-        },
-        onError: function(error) {
-         console.log(error.getErrorMessage());
-        }
-      }).bind("addressBookWidgetDiv");
+          design: {
+            designMode: 'responsive'
+          },
+          onError: function(error) {
+           console.log(error.getErrorMessage());
+          }
+        }).bind("addressBookWidgetDiv");
+      <% else %>
+        new OffAmazonPayments.Widgets.AddressBook({
+          sellerId: '<%= @amazon_gateway.preferred_merchant_id %>',
+          agreementType: 'orderreference',
+          amazonOrderReferenceId: '<%= current_order.amazon_order_reference_id %>',
+          // amazonOrderReferenceId obtained from Address widget
+          displayMode: "Read",
+          design: {
+            designMode: 'responsive'
+          },
+          onError: function(error) {
+           console.log(error.getErrorMessage());
+          }
+        }).bind("addressBookWidgetDiv");
```

Comment:
I have a feeling there's quite a lot of repetitive code in this `AddressBook`. Ideally, there shouldn't be any code repetition.

It's also a bit hard to understand what that if condition with the `try(:retry)` means and does. If needed, wrap it in a class or method to better describe its function.